singer/src/main/java/com/pinterest/singer/writer/s3/S3Writer.java

```diff
@@ -0,0 +1,438 @@
+        }
+    }
+
+    /**
+     * Writes a log message to the buffer file for the current commit.
+     *
+     * @param logMessageAndPosition the log message and its position
+     * @param isDraining whether the system is in a draining state
+     * @throws LogStreamWriterException if an error occurs while writing the log message
+     */
+    @Override
+    public synchronized void writeLogMessageToCommit(LogMessageAndPosition logMessageAndPosition,
+                                                     boolean isDraining) throws LogStreamWriterException {
+        try {
+            writeMessageToBuffer(logMessageAndPosition);
+            messageCount++;
+        } catch (IOException e) {
+            try {
+                bufferedOutputStream = new BufferedOutputStream(new FileOutputStream(bufferFile, true));
+                writeMessageToBuffer(logMessageAndPosition);
+            } catch (IOException ex) {
+                LOG.error("Failed to close buffer writer", ex);
+                throw new LogStreamWriterException("Failed to write log message to commit", e);
+            }
+        }
+    }
+
+    private void writeMessageToBuffer(LogMessageAndPosition logMessageAndPosition) throws IOException {
+        byte[] logMessageBytes = logMessageAndPosition.logMessage.getMessage();
+        bufferedOutputStream.write(logMessageBytes);
+        bufferedOutputStream.write('\n'); // Add a newline character after each log message
+        bufferedOutputStream.flush();
+    }
+
+    /**
+     * Helper function to get the remaining part of the host name after the cluster prefix, typically a UUID.
+     * */
+    public static String extractHostSuffix(String inputStr) {
+        String[] parts = inputStr.split("-");
+        return parts[parts.length - 1];
+    }
+
+    /**
+     * Helper function to get the file name without the .log extension.
+     * */
+    public static String getFileName(String input) {
+        if (input.endsWith(".log")) {
+            return input.substring(0, input.length() - 4);
+        }
+        return input;
+    }
+
+    /**
+     * Gets the actual file name format for the S3 file.
+     * */
+    public String getS3FileFormat() {
+        // FORMAT: <log_name>/<service_fleet>/<host>/<log_dir>/<custom_filename>.<timestamp>
+        String logName = logStream.getSingerLog().getSingerLogConfig().getName();
+        List<String> hostPrefixes = getHostnamePrefixes("-");
+        String serviceFleet = hostPrefixes.get(hostPrefixes.size() - 2);
+        String host = extractHostSuffix(HOSTNAME);
+        String logDir = logStream.getFullPathPrefix().substring(1);
+        String customFilename = s3WriterConfig.getFileNameFormat();
+        String timestamp = formatter.format(new Date());
+        LOG.info("Uploading the file: " + logName + "/" + serviceFleet + "/" + host + "/" + logDir + "/" + customFilename +
```

Comment:
Can we store this in a string instead? So that we can use it for both the log statement and return value

Reply:
ack, I have saved it as a String